lib/rust/ensogl/core/src/control/io/mouse/event.rs

```diff
@@ -236,6 +249,79 @@ macro_rules! define_events {
     }};
 }
 
+/// An JS event type that is convertible to a light copyable struct containing its associated data.
+#[allow(missing_docs)]
+pub trait ToEventData {
+    type Data: Default + Debug + Copy;
+    fn to_data(&self, shape: Shape) -> Self::Data;
+}
+
+/// The data associated with a mouse event. In web environment, it is derived from the mouse event
+/// itself. In test environment, it is passed as a parameter in order to simulate a particular
+/// mouse event.
```

Comment:
Could we prevent testing from complicating production code by adding support for real (mock) data to `web::binding::mock::MouseEvent`, instead of adding this new abstraction?

Reply:
I tried. The mock types right now are all assuming they are ZST and free to transmute between each other, which is a pain to deal with. Especially given that they emulate the inheritance that way. I think there is a way to do it using composition and derefs instead, but that seemeed to me like a hard way to do it. The simple copyable structs per event are definitely easier to deal with, and also simplify the accessors quite a lot (no need to go through JS function call every time you want event position etc.).